### PR TITLE
ISPN-8207 Update eviction tests and add off-heap eviction listener

### PR DESCRIPTION
**Pull request by vjuranek**
https://issues.jboss.org/browse/ISPN-8207
https://issues.jboss.org/browse/ISPN-8216

### CODE REVIEW

File: core/src/test/java/org/infinispan/eviction/impl/OffHeapEvictionFunctionalTest.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package org.infinispan.eviction.impl;
+
+import org.infinispan.configuration.cache.StorageType;
+import org.testng.annotations.Test;
+
+import jdk.nashorn.internal.ir.annotations.Ignore;
```
Comment (tristantarrant) [inline core/src/test/java/org/infinispan/eviction/impl/OffHeapEvictionFunctionalTest.java]: Checkstyle is failing because of unused imports

**Comment by vjuranek** [comment]
ops, sorry, fixed

File: core/src/test/java/org/infinispan/eviction/impl/ObjectEvictionFunctionalTest.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * @author vjuranek
+ * @since 9.2
+ */
+@Test(groups = {"functional", "smoke"}, testName = "eviction.ObjectEvictionFunctionalTest")
+public class ObjectEvictionFunctionalTest extends BaseEvictionFunctionalTest {
+
+   @Override
+   protected StorageType getStorageType() {
+      return StorageType.OBJECT;
```
Comment (wburns) [inline core/src/test/java/org/infinispan/eviction/impl/ObjectEvictionFunctionalTest.java]: We should be able to make this into a `@Factory` method similar to https://github.com/infinispan/infinispan/blob/master/core/src/test/java/org/infinispan/container/offheap/OffHeapSingleNodeExpirationEvictionTest.java#L45.

File: core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java [hunk 3]
```diff
@@ -223,7 +224,9 @@ private void ensureSize() {
                      addressToRemove, currentSize, maxSize);
             }
             try {
+               InternalCacheEntry<WrappedBytes, WrappedBytes> ice = offHeapEntryFactory.fromMemory(addressToRemove);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java]: I wonder if we could change this so we lazily ressurect this object. This way if no listener is installed we wouldn't have to read this object. But that would probably require changing `evictionManager` API. So maybe for ISPN 10? @tristantarrant WDYT?

**Comment by vjuranek** [comment]
@wburns thanks, switched to factory method.

File: core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java (hunk 3 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java]: Sure.

**Comment by tristantarrant** [comment]
Merged to master, thanks

**Pull request merged**